dnf/transaction_sr.py

```diff
@@ -0,0 +1,104 @@
+# Copyright (C) 2020 Red Hat, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Library General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import libdnf
+
+from dnf.i18n import _
+import dnf.exceptions
+
+import json
+
+
+VERSION_MAJOR = "1"
+VERSION_MINOR = "0"
+VERSION = "%s.%s" % (VERSION_MAJOR, VERSION_MINOR)
+
+
+class StoredTransactionVersionError(dnf.exceptions.Error):
+    def __init__(self, value=None):
+        super(StoredTransactionVersionError, self).__init__(value)
+
+
+def check_version(version, filename):
+    major, minor = version.split('.')
+    if major != VERSION_MAJOR:
+        raise StoredTransactionVersionError(
+            _('Incompatible major version {major} in "{fn}", '
+                'supported major version is {major_supp}.'
+            ).format(major=major, fn=filename, major_supp=VERSION_MAJOR)
+        )
+
+    if minor < VERSION_MINOR:
+        raise StoredTransactionVersionError(
+            _('Incompatible minor version {minor} in "{fn}", '
+                'supported minor version is {minor_supp} or greater.'
+            ).format(minor=minor, fn=filename, minor_supp=VERSION_MINOR)
+        )
+
+
+def store_transaction(transaction, filename="transaction.json"):
+    data = {
+        "version": VERSION,
+        "solver_flags": "",  # TODO
+    }
+    rpms = {}
+    groups = {}
+    environments = {}
+
+    for i in transaction.packages():
+        if i.is_package():
+            data_p = rpms.setdefault("%s.%s" % (i.name, i.arch), {})
+            if i._item.isForwardAction():
+                data_p["to"] = i.evr
+                data_p["action"] = i.action_name
+            elif i._item.isBackwardAction():
+                data_p["from"] = i.evr
+                # make sure not to overwrite the forward action if it was written already
+                if "action" not in data_p:
+                    data_p["action"] = i.action_name
+            else:
+                data_p["weird_reason"] = True
+
+            data_p["reason"] = libdnf.transaction.TransactionItemReasonToString(i.reason)
+        elif i.is_group():
+            group = i.get_group()
+
+            data_p = groups.setdefault(group.getGroupId(), {})
```

Comment:
What happen when I will use the shell command to install and remove the same group?

Reply:
> What happen when I will use the shell command to install and remove the same group?

Surprisingly enough, doing that will create a transaction where the group is installed even though it was removed in the shell before running the transaction. It would appear that's a bug?

Reply:
Not sure.

Reply:
What about to also use a list instead of dict? How can we be sure that there will be not multiple actions with the same group/env_group id? Again small change like for rpms but it could prevents some issues.

Reply:
I'd like to say that we better be sure there aren't multiple operations on the same group just so that our code actually maintains some level of sanity.